Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp

```diff
@@ -0,0 +1,448 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Core/HW/HSP/HSP_DeviceGBPlayer.h"
+
+#include "mgba/flags.h"
+
+#include "mgba-util/vfs.h"
+#include "mgba/core/blip_buf.h"
+#include "mgba/core/core.h"
+#include "mgba/core/log.h"
+#include "mgba/gba/core.h"
+#include "mgba/gba/interface.h"
+
+#include "Common/ChunkFile.h"
+#include "Common/Logging/Log.h"
+#include "Common/MemoryUtil.h"
+#include "Common/Thread.h"
+
+#include "Core/ConfigManager.h"
+#include "Core/CoreTiming.h"
+#include "Core/HW/DSP.h"
+#include "Core/HW/HSP/HSP.h"
+#include "Core/HW/ProcessorInterface.h"
+
+#include <condition_variable>
+#include <cstring>
+#include <mutex>
+
+namespace HSP
+{
+enum GBPRegister
+{
+  GBP_TEST = 0x10,
+  GBP_VIDEO = 0x11,
+  GBP_CONTROL = 0x14,
+  GBP_SIO_CONTROL = 0x15,
+  GBP_AUDIO = 0x18,
+  GBP_SIO = 0x19,
+  GBP_KEYPAD = 0x1C,
+  GBP_IRQ = 0x1D
+};
+
+enum
+{
+  IRQ_ASSERTED = 0x8000
+};
+
+enum
+{
+  CONTROL_CART_DETECTED = 0x01,
+  CONTROL_CART_INSERTED = 0x02,
+  CONTROL_3V = 0x04,
+  CONTROL_5V = 0x08,
+  CONTROL_MASK_IRQ = 0x10,
+  CONTROL_SLEEP = 0x20,
+  CONTROL_LINK_CABLE = 0x40,
+  CONTROL_LINK_ENABLE = 0x80
+};
+
+enum class CHSPDevice_GBPlayer::IRQ : int
+{
+  LINK = 0,
+  GAME_PAK = 1,
+  SLEEP = 2,
+  SERIAL = 3,
+  VIDEO = 4,
+  AUDIO = 5
+};
+
+class CGBPlayer_mGBA : public IGBPlayer
+{
+public:
+  CGBPlayer_mGBA(CHSPDevice_GBPlayer*);
+  virtual ~CGBPlayer_mGBA();
+
+  virtual void Reset();
+  virtual void Shutdown();
+
+  virtual bool IsLoaded() const;
+  virtual bool IsGBA() const;
+
+  virtual bool LoadGame(const std::string& path);
+  virtual bool LoadBootrom(const std::string& path);
+
+  virtual void ReadScanlines(u32* scanlines);
+  virtual void ReadAudio(u8* audio);
+
+  virtual void SetKeys(u16 keys);
+
+private:
+  mCore* m_core = nullptr;
+  u32* m_framebuffer;
+  CoreTiming::EventType* m_gbp_update_event = nullptr;
+
+  bool m_scanlines_ready = false;
+  int m_next_scanlines = 0;
+  int m_next_audio = 0;
+  u16 m_keys = 0;
+  s16 m_audio[2][4096];
+
+  mLogger s_logger;
+
+  std::thread m_thread;
+  std::mutex m_mutex;
+  std::condition_variable m_cond;
+
+  static void Update(u64 user_data, s64 cycles_late);
+  void Update(s64 cycles_late);
+
+  void GBAThread();
+};
+
+CHSPDevice_GBPlayer::CHSPDevice_GBPlayer(HSPDevices device)
+    : IHSPDevice(device), m_test{}, m_control(0xCC), m_irq(0x0000)
+{
+  memset(m_scanlines, 0x00, sizeof(m_scanlines));
```

Comment:
This can be removed by just adding a `= {};` to the declaration in the class body.